Emit an Audit log event when a Bot or Instance fails to join

strideynet opened the pull request:
Currently, a failed join is not audit logged or even logged. This makes it really hard to determine why a join failed or to detect a malicious join attempt. This PR introduces an audit log event which is emitted when a join fails for any reason.

For join methods that support returning the attributes of the joining identity, these will also show in the audit log. Not all join methods support this yet, and we ought to return to those and add the appropriate methods so they do surface this information which is extremely useful in diagnosing problems with joining and allow rules.

![image](https://github.com/gravitational/teleport/assets/16336790/0a966a83-6220-4770-91ab-7d765df25e42)

![image](https://github.com/gravitational/teleport/assets/16336790/c08684c3-0688-432f-ab9d-5763cb2950d2)

changelog: Adds a new Audit log event that is emitted when an Agent or Bot request to join the cluster is denied.

Closes https://github.com/gravitational/teleport/issues/17948

strideynet committed:
Emit audit event on instance/bot join failure

strideynet committed:
Add audit log on failure to azure and iam join methods

strideynet committed:
Tidy up a lil bit

strideynet committed:
Add missing icon to EventTypeCell.tsx

strideynet committed:
Regenerate snapshot

avatus commented on lib/auth/join.go:
```diff
@@ -114,6 +115,92 @@ func setRemoteAddrFromContext(ctx context.Context, req *types.RegisterUsingToken
+				Code: events.BotJoinFailureCode,
+			},
+			Status:     status,
+			BotName:    "unknown",
+			Method:     "unknown",
```
is it our convention to add `"unknown"` in all of these fields instead of leaving them empty?

avatus replied:
Hm, I'm unsure as well. To me, `"unknown"` throws a much bigger red flag than empty but I could be wrong. If we've already implemented other unknowns, then perhaps we shouldn't change it here until we have a better idea/plan. If it's a UI thing, we could bring it up at the UX/UI open hours to see if they have a better solution?

ibeckermayer replied:
It looks like I did something similar in various desktop events e.g.

https://github.com/gravitational/teleport/blob/39882ea520705b8a6ff8114a5b7653117a1a81de/lib/srv/desktop/audit.go#L267

It could make sense to standardize this along the lines of:

- `""`: consider this a bug
- `"unknown"`: typically events of this kind would have this field, but it could not be known at the time of event emission

The idea being that we want to be able to clearly distinguish between data loss due to bugs (which are high priority fixes, due to the security-relevance of audit events) vs data which we explicitly omit by necessity.

However I notice that e.g `BotName` is `"omitempty"`

https://github.com/gravitational/teleport/blob/39882ea520705b8a6ff8114a5b7653117a1a81de/api/proto/teleport/legacy/types/events/events.proto#L3740-L3741

so this particular case might be as @avatus suggested, just worthy of a client side display fix.

strideynet replied:
> It could make sense to standardize this along the lines of:
> 
> - "": consider this a bug
> - "unknown": typically events of this kind would have this field, but it could not be known at the time of event emission

Imho this is a good direction to head in - audit logs contain important information and when this information is missing, we need to act. Distinguishing between audit logs that have been emitted in a place where a certain field is not yet known, and those that are missing the information because of a bug or because they predate the field having been added is useful.

> However I notice that e.g BotName is "omitempty"

I'm not too sure this has a significant impact, at the time when I added these events, I wasn't too sure whether to use omitempty or not.

strideynet replied:
After a chat with Edoardo, I'm going to remove these "unknown" default values for now.

ibeckermayer replied:
tl;dr of your chat out of curiousity?

ibeckermayer commented on lib/events/codes.go:
```diff
@@ -420,8 +420,12 @@ const (
 	// InstanceJoinCode is the 'node.join' event code.
 	InstanceJoinCode = "TJ002I"
+	// InstanceJoinFailureCode is the 'node.join' event code for failures.
+	InstanceJoinFailureCode = "TJ001E"
```
```suggestion
	InstanceJoinFailureCode = "TJ002E"
```

strideynet replied:
243bd63ec5

strideynet commented on lib/auth/join.go:
```diff
@@ -144,40 +239,52 @@ func (a *Server) RegisterUsingToken(ctx context.Context, req *types.RegisterUsin
 			"sure your node is configured to use the %s join method", method, method)
 	case types.JoinMethodGitHub:
 		claims, err := a.checkGitHubJoinRequest(ctx, req)
```
I think you might be mistaken - the behaviour of named returns in Go is a little special. Even if you redefine the variable name is a lower scope, the value that is actually returned is "put" into the named variable when executing defer.

https://go.dev/play/p/VZfeCQi7TdH

ibeckermayer replied:
I stand corrected, thanks!

strideynet committed:
Correct status code

strideynet committed:
Merge remote-tracking branch 'origin/master' into strideynet/audit-join-failures

strideynet committed:
Add host_id and node_name to the join audit log

strideynet committed:
Remove "unknown" default values

strideynet commented:
@rosstimothy 

> https://github.com/gravitational/teleport/pull/40729 added some logging for failed login attempts. Are we now going to log things twice? Does this take any of those changes into account?

I've already resolved the conflicts with this PR, I've removed those log lines in favour of my ones, and added any details that were missing from mine. I feel it still makes sense to move ahead with my PR since 40729 did not add an audit log event, and for our cloud customers, they do not have access to the logs of Teleport. The audit log is therefore the only way we can provide feedback about misconfigured joins.

I'll hold off merging until you've had a chance to see/respond to this :)
